src/lib/common/sol-mainloop-common.c

```diff
@@ -296,6 +347,177 @@ sol_mainloop_common_idler_first(void)
     return NULL;
 }
 
+/* must be called with mainloop lock HELD */
+static inline void
+source_cleanup(void)
+{
+    struct sol_mainloop_source_common *source;
+    uint16_t i;
+
+    if (!source_pending_deletion)
+        return;
+
+    // Walk backwards so deletion doesn't impact the indices.
```

Comment:
Every single use of backwards-loop with sol-vector has this comment. I wonder if something akin to C++'s `std::remove_if` would be feasible with some macro trickery.